### PR TITLE
Preserve aggregated cluster roles on upgrade

### PR DESCRIPTION
**Pull request by enj**
This change adds admin, edit and view to the ClusterRolesToAggregate field in our bootstrap PolicyData.  This guarantees that reconciliation will correctly preserve any changes made to these cluster roles before aggregation was enabled.  Added a unit test to track aggregation based on the bootstrap PolicyData so changes are more obvious in the future.

[Bug 1537426](https://bugzilla.redhat.com/show_bug.cgi?id=1537426)

Signed-off-by: Monis Khan <mkhan@redhat.com>

/assign @deads2k @simo5
/kind bug

### CODE REVIEW

**Comment by enj** [comment]
@openshift/sig-security

**Comment by deads2k** [comment]
@ingvagabund concrete example of "compatible" API (just added an optional field) effectively breaking the API because of how it is used.

Even worse, in this case it was my upstream change that made it and my downstream change (quite some time later) that missed it.

File: pkg/cmd/server/bootstrappolicy/all_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	// check that the new name of the aggregated cluster role is valid
+	// aggregationMap is effectively old name -> all possible valid new names
+	// clusterRolesToAggregate is old name -> new name
+	for parentClusterRole, childClusterRoles := range aggregationMap {
+		newNameOfClusterRole, ok := clusterRolesToAggregate[parentClusterRole]
+		// we would have caught this earlier but does not hurt to check again
+		if !ok {
+			t.Errorf("cluster role %s in missing from %#v", parentClusterRole, clusterRolesToAggregate)
```
Comment (php-coder) [inline pkg/cmd/server/bootstrappolicy/all_test.go]: s/in missing/is missing/

File: pkg/cmd/server/bootstrappolicy/all_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"strings"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const osClusterRoleAggregationPrefix = "system:openshift:"
```
Comment (php-coder) [inline pkg/cmd/server/bootstrappolicy/all_test.go]: Do we really need this constant if we use it only once? 5 lines bellow we don't use it, for example.

File: pkg/cmd/server/bootstrappolicy/all_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+func TestPolicyAggregation(t *testing.T) {
+	policyData := Policy()
+	clusterRoles := policyData.ClusterRoles
+	clusterRolesToAggregate := policyData.ClusterRolesToAggregate
+
+	// do some basic sanity checks
+
+	if len(clusterRoles) == 0 || len(clusterRolesToAggregate) == 0 {
+		t.Fatalf("invalid policy data:\n%#v\n%#v", clusterRoles, clusterRolesToAggregate)
```
Comment (php-coder) [inline pkg/cmd/server/bootstrappolicy/all_test.go]: I'd expect that the error message also mention what is the expected state.

File: pkg/cmd/server/bootstrappolicy/all_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+		hasAggregationRuleSet.Insert(cr.Name)
+
+		// insert this cluster role into aggregationMap with all of cluster roles that aggregate into it
+		for j := range cr.AggregationRule.ClusterRoleSelectors {
+			labelSelector := cr.AggregationRule.ClusterRoleSelectors[j]
+			selector, err := v1.LabelSelectorAsSelector(&labelSelector)
+			if err != nil {
+				// should never happen
+				t.Errorf("invalid label selector %#v   at index %d for cluster role %s: %v", labelSelector, j, cr.Name, err)
```
Comment (php-coder) [inline pkg/cmd/server/bootstrappolicy/all_test.go]: are these extra spaces in the message (`%#v   at index`) were added intentionally?

File: pkg/cmd/server/bootstrappolicy/all_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+// this map must be manually kept up to date as we make changes to aggregation
+// we hard code this data with no constants because we cannot change the underlying values
+var expectedAggregationMap = map[string]sets.String{
+	"admin": sets.NewString("system:openshift:aggregate-to-admin", "system:aggregate-to-admin"),
+	"edit":  sets.NewString("system:openshift:aggregate-to-edit", "system:aggregate-to-edit"),
+	"view":  sets.NewString("system:openshift:aggregate-to-view", "system:aggregate-to-view"),
+}
+
+func TestPolicyAggregation(t *testing.T) {
```
Comment (php-coder) [inline pkg/cmd/server/bootstrappolicy/all_test.go]: The test has ±100 lines, 3 nested `for` loops and many statements that do something. Unfortunately, the testing logic isn't obvious and I'm curious, could the test be improved? Even the fact that you have to put some comments is a hint that probably some functions could be extracted.

**Comment by ingvagabund** [comment]
@deads2k would be valuable to at least report all the places where a struct is initialized and a new version/commit introduces new field?

**Comment by deads2k** [comment]
> @deads2k would be valuable to at least report all the places where a struct is initialized and a new version/commit introduces new field?

Depends on the goal of the tool.  I was thinking about from the perspective of automatically bumping semantic versions when required, so the original change would have bumped a minor level and callers would be accidentally broken.

Thinking out loud, I'm not sure what I'd do with that information.  Without knowing which structs are broken ahead of time, being able to track the ones we create isn't very useful.

**Pull request merged**